nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+
+```javascript
+// asyncHandler lets us wrap async express routes to handle errors.
+asyncHandler(async (req, res, next) => {
+  const errors = validationResult(req);
+  
+  if (!errors.isEmpty()) {
+    return res.status(400).json({ errors: errors.array() });
+  } else {
+    res.send("Data is valid, form submitted!");
+  }
+});
+```
+
+This setup checks for any failed validation checks, and if there are any (the errors array is NOT empty), then it sends a 400 code, along with any errors that may be present.
+
+### Forms and Express routes
+
+One final thing to cover is how to handle routes in Express. After all, our form needs somewhere to send the data to.
+
+Since we don't want our routes handling logic, we instead use a controller.
+
+```javascript
+const asyncHandler = require("express-async-handler");
+
+exports.user_update_get = asyncHandler(async (req, res, next) => {});
+exports.user_update_post = asyncHandler(async (req, res, next) => {});
+```
+
+Inside our router, we can then assign routes which correspond to the controller's functions:
+
+```javascript
+const { Router } = require("express");
+const usersRouter = Router();
+const usersController = require("../controllers/usersController");
+
+// User update routes
+usersRouter.get("/users/:id/update", usersController.user_update_get);
+usersRouter.post("/users/:id/update", usersController.user_update_post);
+
+module.exports = usersRouter;
+```
+
+You'll notice there's a respective `GET` and `POST` route.
+
+- `GET` routes typically display forms or confirmation pages.
+- `POST` routes handle form submissions and data processing.
+
+You may also notice the `:id` parameter in the route. This lets us pass a value to those routes that we can use to tell the controller which `User` object we're referencing.
+
+And in our form, the action would look something like this:
+
+```ejs
+<!-- Example using EJS with POST to submit an update to our Express server. -->
+<form action="/users/<%= user.userId %>/update" method="POST"></form>
+```
+
+`/users/:id/update` is an *endpoint* we've created on our Express server.
+
+### Handling errors
+
+What happens if we don't give our route a valid `:id` to delete or update? We need a way to tell our users an error occurred.
+
+```javascript
+// Handles POST requests for the /users/:id/delete route.
+// Status code 400 can indicate malformed request syntax, or a missing :id in our case.
+exports.user_delete_post = asyncHandler(async (req, res, next) => {
+  if (!req.params.id) {
+    return res.status(400).send("No user id provided.");
+  }
+});
+```
+
+By responding with a status code and a message, we give the client a way to both handle the error code, and a way to display an appropriate error message. We could also render a new or updated view which we'll see below.
```

Comment:
Not sure if the lesson needs to handle params error handling here. What do you think?

Reply:
I have no strong opinion on it. I think it's fine to remove since the `### Validation results``` covers handling validation errors. Just lmk and I'll take it out.

Reply:
Yeah, let's remove this.